### PR TITLE
Improve jprod_residual! and jtprod_residual!

### PR DESCRIPTION
**Pull request by amontoison**

### CODE REVIEW

File: src/moi_nls_model.jl [hunk 1]
```diff
@@ -143,67 +143,43 @@ end
...
+  increment!(nls, :neval_jprod_residual)
+  if nls.nls_meta.nnln > 0
+    MOI.eval_constraint_jacobian_product(nls.Feval, Jv, x, v)
+  end
+  if nls.nls_meta.nlin > 0
+    nls.nls_meta.nnln == 0 && (Jv .= 0.0)
+    for k = 1:nls.linequ.nnzj
+      row, col, val = nls.linequ.jacobian.rows[k], nls.linequ.jacobian.cols[k], nls.linequ.jacobian.vals[k]
+      Jv[row] += v[col] * val
+    end
```
Comment (tmigot) [inline src/moi_nls_model.jl:158]:
```suggestion
nls.nls_meta.nlin > 0 && (Jv .= 0.0)
  if nls.nls_meta.nnln > 0
    MOI.eval_constraint_jacobian_product(nls.Feval, view(Jv, nls.nls_meta.nln), x, v)
  end
  if nls.nls_meta.nlin > 0
    for k = 1:nls.linequ.nnzj
      row, col, val = nls.linequ.jacobian.rows[k], nls.linequ.jacobian.cols[k], nls.linequ.jacobian.vals[k]
      Jv[row] += v[col] * val
    end
  end
```

File: src/moi_nls_model.jl [hunk 2]
```diff
@@ -143,67 +143,43 @@ end
...
+  if nls.nls_meta.nnln > 0
+    MOI.eval_constraint_jacobian_transpose_product(nls.Feval, Jtv, x, v)
+  end
+  if nls.nls_meta.nlin > 0
+    nls.nls_meta.nnln == 0 && (Jtv .= 0.0)
+    for k = 1:nls.linequ.nnzj
+      row, col, val = nls.linequ.jacobian.rows[k], nls.linequ.jacobian.cols[k], nls.linequ.jacobian.vals[k]
+      Jtv[col] += v[row] * val
+    end
+  end
```
Comment (tmigot) [inline src/moi_nls_model.jl:179]:
```suggestion
nls.nls_meta.nlin > 0 && (Jtv .= 0.0)
  if nls.nls_meta.nnln > 0
    MOI.eval_constraint_jacobian_transpose_product(nls.Feval, Jtv, x, view(v, nls.nls_meta.nln))
  end
  if nls.nls_meta.nlin > 0
    for k = 1:nls.linequ.nnzj
      row, col, val = nls.linequ.jacobian.rows[k], nls.linequ.jacobian.cols[k], nls.linequ.jacobian.vals[k]
      Jtv[col] += v[row] * val
    end
  end
```

**Review by tmigot — CHANGES_REQUESTED** [review]
Here are some suggestions.

Are the indices of the linear jacobian of the residual in relative indices or real indices?

File: src/moi_nls_model.jl [hunk 3]
```diff
@@ -143,67 +143,44 @@ end
...
-function NLPModels.jtprod_residual!(
-  nls::MathOptNLSModel,
-  x::AbstractVector,
-  v::AbstractVector,
-  Jtv::AbstractVector,
-)
-  rows, cols = jac_structure_residual(nls)
-  jtprod_residual!(nls, x, rows, cols, v, Jtv)
+  increment!(nls, :neval_jtprod_residual)
+nls.nls_meta.nlin > 0 && (Jtv .= 0.0)
```
Comment (amontoison) [inline src/moi_nls_model.jl:171]:
```suggestion
  nls.nls_meta.nlin > 0 && (Jtv .= 0.0)
```

File: src/moi_nls_model.jl [hunk 4]
```diff
@@ -143,67 +143,44 @@ end
...
-function NLPModels.jprod_residual!(
-  nls::MathOptNLSModel,
-  x::AbstractVector,
-  v::AbstractVector,
-  Jv::AbstractVector,
-)
-  rows, cols = jac_structure_residual(nls)
-  jprod_residual!(nls, x, rows, cols, v, Jv)
+  increment!(nls, :neval_jprod_residual)
+nls.nls_meta.nlin > 0 && (Jv .= 0.0)
```
Comment (amontoison) [inline src/moi_nls_model.jl:150]:
```suggestion
  nls.nls_meta.nlin > 0 && (Jv .= 0.0)
```

**Comment by amontoison** [comment]
> Here are some suggestions.
> 
> Are the indices of the linear jacobian of the residual in relative indices or real indices?

Thanks Tangi! 
The indices of the linear jacobian of the residual are real indices.

**Pull request merged**